app/utils/visualizations/crop_cloud.py

```diff
@@ -0,0 +1,659 @@
+
+    # add complexity
+    user_words["complexity"] = get_complexity(user_words, metric=complexity_metric)
+    user_words.sort_values(by='complexity', ascending=False, inplace=True)
+    return user_words
+
+# Map the dates to integers. returns a dictionary
+def map_dates(user_words):
+    dates = sorted(user_words.date.unique())
+    D = len(dates)
+    date_map = dict()
+    for i, date in enumerate(dates):
+        date = pd.to_datetime(date)
+        date_map[date] = i
+    return date_map
+
+# Picks a random horizontal location for a cropped word
+# This is the heart of arranging the words chronologically
+def pick_x(canvas_width, word_width, date_number=None, total_dates=None):
+    if total_dates:  # this is untested and may need to be debugged
+        # for now, this divides the space into even fractions
+        # I had wanted to use arc cosine waves to give a fuzzy distribution, but never finished the function bending
+        x_float = np.random.uniform(
+            low=date_number/total_dates,
+            high=(date_number+1)/total_dates,
+            )
+    else:
+        x_float = np.random.uniform()
+    available_room = canvas_width - word_width
+    return int(x_float * available_room)
+
+# Picks a random vertical location for a cropped word
+# This uses a triangular distribution which biases the words towards the midline, where your eyes will start
+def pick_y(canvas_height, word_height):
+    y_float = np.random.triangular(0, 0.5, 1)
+    available_room = canvas_height - word_height
+    return int(y_float * available_room)
+
+# Constructs and renders a crop cloud. Returns an image
+def make_crop_cloud(canvas, boxes):
+    # propose a location for the word
+    # does this location collide with anything already placed?
+    # if not, then place the word
+    # OpenCV uses [y:x] coordinates for images
+
+    occupied = np.zeros(shape=(canvas.shape[:2]), dtype=bool)
+    placed = 0
+    collisions = 0
+    total = len(boxes)
+    word_area = 0
+    for row in boxes.to_dict('records'):
+        image = row['image']
+
+        max_attempts = 20
+        failed_attempts = 0
+        while failed_attempts < max_attempts:
+            # pick a horizontal position
+            x1 = pick_x(
+                canvas_width=canvas.shape[1],
+                word_width=image.shape[1],
+                date_number=0,
+                total_dates=1,
+                )
+            x2 = x1 + image.shape[1]
+
+            # pick a vertical position
+            y1 = pick_y(canvas_height=canvas.shape[0], word_height=image.shape[0])
+            y2 = y1 + image.shape[0]
+
+            R, G, B, mask = cv2.split(image)  # split the image into channels
+            color = cv2.merge((B, G, R))
+            mask = np.atleast_3d(mask) / 255
+
+            mask_bool = mask.reshape(mask.shape[:2]).astype('bool')
+            intersection = np.logical_and(mask_bool, occupied[y1:y2, x1:x2])
+
```

Comment:
I've said it before, and I'll say it again: really impressed by the from scratch crop cloud code here. Im guessing Ill implement color changes somewhere around here....